sql,kv: replace InitPuts with CPuts

yuzefovich opened the pull request:
**sql: replace InitPut with CPut**

CPut with empty `expValue` parameter is the same as InitPut (with `failOnTombstones=false`) when we don't expect the key to exist, so we replace the latter with the former throughout the code base. If the key might already exist, then we need to use CPut with the possible value and `AllowIfDoesNotExist` set to `true` (which is the case for the backfill code where the value might exist when performing the primary key change). This paves the way to deprecating the InitPut request altogether.

**kv: extend CPut with FailOnTombstones option**

This commit extends the CPut command with FailOnTombstones option that we already have in the InitPut - it makes it so that the CPut fails with a ConditionFailedError if it encounters a tombstone. This begins the process of folding all usages of InitPuts into CPuts. (We actually have only one place where we use InitPuts with this flag set - in the sql liveness system which allows us to ensure that after a session is dead, it cannot be resurrected, which allows us to keep the dead session cache easily.)

InitPut request was introduced long time ago in 1a956303129f3c8e3adaf0f7f1e53193fb16fba5 when we considered extending CPut with a boolean for special tombstone handling too burdensome and ugly. Since then we have extended the CPut with a few options, so adding an extra one doesn't seem like a big deal anymore. On the plus side, it removes the confusion that I personally had for why we used InitPut as opposed to CPut with nil expected value in a few places (all were removed in the previous commit).

The deprecation was alluded to during the review of #35157.

Release note: None

Epic: None
Addressed: #71074.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/137808)
<!-- Reviewable:end -->

yuzefovich committed:
sql: replace InitPut with CPut

CPut with empty `expValue` parameter is the same as InitPut (with
`failOnTombstones=false`) when we don't expect the key to exist, so we
replace the latter with the former throughout the code base. If the key
might already exist, then we need to use CPut with the possible value
and `AllowIfDoesNotExist` set to `true` (which is the case for the
backfill code where the value might exist when performing the primary
key change). This paves the way to deprecating the InitPut request
altogether.

Release note: None

yuzefovich committed:
kv: extend CPut with FailOnTombstones option

This commit extends the CPut command with FailOnTombstones option that
we already have in the InitPut - it makes it so that the CPut fails with
a ConditionFailedError if it encounters a tombstone. This begins the
process of folding all usages of InitPuts into CPuts. (We actually have
only one place where we use InitPuts with this flag set - in the sql
liveness system which allows us to ensure that after a session is dead,
it cannot be resurrected, which allows us to keep the dead session cache
easily.)

InitPut request was introduced long time ago in
1a956303129f3c8e3adaf0f7f1e53193fb16fba5 when we considered extending
CPut with a boolean for special tombstone handling too burdensome and
ugly. Since then we have extended the CPut with a few options, so
adding an extra one doesn't seem like a big deal anymore. On the plus
side, it removes the confusion that I personally had for why we used
InitPut as opposed to CPut with nil expected value in a few places (all
were removed in the previous commit).

The deprecation was alluded to during the review of #35157.

Release note: None

michae2 commented:
Just wondering: is this a rebase of #88519 or a fresh PR?

yuzefovich commented:
Oh, I wasn't aware of that PR, I did this one from scratch.

yuzefovich commented:
Looks there are two main differences between Nathan's PR and this one:
- in #88519 we used an equivalent transformation of `InitPut` into `CPut` with `AllowIfDoesNotExist=true` and nil `expValue`. In this change I'm replacing almost all `InitPut` usages with a `CPut` with `AllowIfDoesNotExist` not set since my understanding is that we don't expect the key to exist in most cases.
- #88519 attempt to remove `InitPut` request type completely, but since the time that PR was open, we now have a single call site where we do use `InitPut` with `FailOnTombstones` which we need to handle separately. In this PR we begin the process of deprecating client-side usages of `InitPut`.

yuzefovich commented:
I typed up an alternative PR #138707 assuming that we can remove the single spot where we use InitPut with FailOnTombstones option.

yuzefovich commented:
Let's proceed with #138707, so I'll put this into the draft mode for now.

yuzefovich commented:
Superseded by #138707
